src/http/tracker_interface.rs

```diff
@@ -0,0 +1,100 @@
+use std::future::Future;
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use futures::future::BoxFuture;
+
+use crate::signals::shutdown_signal;
+use crate::tracker::Tracker;
+
+/// Trait to be implemented by a http interface for the tracker.
+#[allow(clippy::module_name_repetitions)]
+pub trait TrackerInterfaceTrait: Sync + Send {
```

Comment:
I see you have the other struct:

```rust
pub struct TrackerInterface<S> {
    cfg: torrust_tracker_configuration::HttpTracker,
    state: S,
}
```

Maybe we could call this one just `HttpTrackerHandler` or `HttpServerHandler`, depending on whether we want to use `HttpServer` or `HttpTrackerServer` for the thread running the service.

Reply:
In summary, you would have:

```rust
pub trait HttpServerStarter: Sync + Send { 
    fn start_with_graceful_shutdown<F>(...) -> ...
}

pub struct HttpServer<S> {
    cfg: torrust_tracker_configuration::HttpTracker,
    state: S,
}

pub type StoppedHttpServer<I> = HttpServer<Stopped<I>>;
pub type RunningHttpServer<I> = HttpServer<Running<I>>;

impl<I: HttpServerStarter + 'static> StoppedHttpServer<I> { ... }
impl<I: HttpServerStarter> RunningHttpServer<I> { ... }
```

Reply:
Hi @josecelano ,

To provide a bit of background to my commits:

We have talked about making the tracker agnostic of the `server` or `interface` that is being used to access it, so that we can run it using any `http` or `udp` server. As of right now, we already have two `http` server implementations: `Axum` and `Warp`.

To not have to deal with a separate server struct for every implementation when starting them, I have introduced the `TrackerInterface` wrapper struct. This `TrackerInterface` struct wraps any http server implementation that implements the trait `TrackerInterfaceTrait`. Like [axum](https://github.com/torrust/torrust-tracker/pull/170/files#diff-7d63bb6e132aa19415e0d8153f9fe941dca3c4c4e33c347ef7ffe8b0eeab99e2R80) and [warp](https://github.com/torrust/torrust-tracker/pull/170/files#diff-c7ccce3f8a6fe2f7058e491d032ee885ba8b409b3ddbca1ba671a854d3c9c3dcR53).

The reason that I chose the word `interface` is because to me the `Tracker` struct is the core of our tracker. And `udp` and `http` are just interfaces to interact with the public functions of our core `Tracker`. But I see that the name might be too confusing.

I think that naming the types like this would avoid any confusion:

`struct` `http::TrackerInterface` -> `http::HttpServerWrapper`
`trait` `http::TrackerInterfaceTrait` -> `http::HttpServer`

What do you think?

Reply:
Actually after thinking about it I like the name “HttpServerStarter” :)

As the struct that implements the trait is only a launcher of the a http server and not the entire server itself.

Reply:
hi, @WarmBeer thank you for clarifying it!

First at all, We are going to remove the Warp implementation soon, but I guess we could have different tracker versions in the future: V1, V2, etcetera. But in that case, I think it would not work because we would break that "interface".

Secondly, I agree that the `Tracker` is the "core" tracker, and the UDP and HTTP trackers are wrappers, but I think they are not only thin wrappers. From my perspective, the `Tracker` is a domain service and the `UDPTracker`, and `HTTPTracker` are application services.

Reply:
> Actually after thinking about it I like the name “HttpServerStarter” :)
> 
> As the struct that implements the trait is only a launcher of the a http server and not the entire server itself.

I think its only responsibilities are:

- Launching the actual server
- Knowing its state
- Knowing the configuration needed to launch it. Secondary responsibility derived from the first one.